### ISSUE TITLE
Entity attributes, attribute triplets and custom loss function

### ISSUE THREAD

**Issue by j-linx** [facebookresearch/PyTorch-BigGraph#257]
I have a transE model built on BigGraph with entity types {A, B, C, D} and various relations mapping between them. Now, as I hope to extend the model to factor in entity attributes a la the [TransEA](https://aclanthology.org/W18-3017.pdf) model, I am having difficulties figuring out how to incorporate these node attributes. 

For context and a quick summary, that proposed model very simply incorporates numeric node attributes by adding numeric attributes as additional triplets <entity, attribute, value>, and then appending the standard TransE margin loss function, L_R, with a regression loss, L_A (\sum\norm{a’e+b-v}), for attributive triplets. This also generates attribute embeddings. A parameter alpha then trades off the standard TransE loss with the new attribute loss term, giving the overall loss function (1-alpha)L_R + alpha*L_A.

I see that the [EntitySchema](https://github.com/facebookresearch/PyTorch-BigGraph/blob/756615d6bc3916b78b5d1f212dfbc194ce07e145/torchbiggraph/config.py#L55) does not allow for an associated attribute (numeric or otherwise). 

I also considered inputting the attributes as relations and the attribute values (bucketed into bins) as the RHS of the triplets, using an affine operator (`'operator': 'affine':`). However, I fundamentally don’t see a way to properly incorporate the updated loss term to reflect what is needed here. Even if I wrote a custom torch loss function I’m not sure how I would then tell BigGraph to only apply the regression loss term to triplets corresponding to the attribute relation and apply the standard loss term to all other triplets, much less incorporate a parameter alpha to weight these loss terms.

I also tried framing the attributes as featurized entities by setting `featurized=True` however since each entity will only have attribute in relation to one RHS attribute value bin, this does not seem right. This feature appears almost entirely geared towards an NLP use case.

Is this type of change, converting TransE to TransEA with a modified loss function, simply not possible under the current framework? Any intuition or guidance on the extensibility potential would be welcome.

**Comment by adamlerer** [comment]
I don't think this is straightforward to implement under the current framework. We've considered adding dense node features in the past but it hasn't been a highly requested feature.
